### PR TITLE
Lock: Forward user cancellation token downstream

### PR DESCRIPTION
**Pull request by mfkl**
This is based on https://github.com/G-Research/consuldotnet/pull/66 for now, otherwise it throws the null ref as well. 

So once https://github.com/G-Research/consuldotnet/pull/66 is merged, I will rebase and cleanup this.

Partial fix for https://github.com/G-Research/consuldotnet/issues/65

### CODE REVIEW

File: Consul.Test/SemaphoreTest.cs [hunk 1]
```diff
@@ -363,5 +365,110 @@ public async Task Semaphore_Conflict()
...
+            {
+                await distributedLock.Acquire(cts.Token);
+            }
+            catch (Exception) { }
+            cancellationOperationTimer.Stop();
+
+            // Assert
+            var stopTimeMs = cancellationOperationTimer.ElapsedMilliseconds;
+            var LockWaitTimeMs = TimeSpan.FromSeconds(15).TotalMilliseconds;
+            Assert.True(stopTimeMs < LockWaitTimeMs);
```
Comment (mfkl) [inline Consul.Test/SemaphoreTest.cs:463]: I don't like this way of asserting, though it does the job. Alternatives welcome.

Comment (marcin-krystianc) [reply Consul.Test/SemaphoreTest.cs:463]: I agree, the NUnit3 approach seems to be much more expressive (e.g. https://moleseyhill.com/2018-12-01-nunit-assert-that-examples.html), but I don't think that mixing XUnit and NUnit assertions would be a good idea.

**Comment by mfkl** [comment]
Let's merge this? @marcin-krystianc @jgiannuzzi

**Comment by jgiannuzzi** [comment]
@mfkl could you please rebase on master now that #66 has been merged?

**Comment by mfkl** [comment]
Done

File: Consul.Test/SemaphoreTest.cs [hunk 2]
```diff
@@ -417,5 +417,54 @@ public async Task Cancelling_A_Token_When_Acquiring_A_Lock_Should_Throw_LockNotH
...
+            try
+            {
+                await distributedLock.Acquire(cts.Token);
+            }
+            catch (Exception) { }
+            cancellationOperationTimer.Stop();
+
+            // Assert
+            var stopTimeMs = cancellationOperationTimer.ElapsedMilliseconds;
+            var LockWaitTimeMs = TimeSpan.FromSeconds(15).TotalMilliseconds;
```
Comment (marcin-krystianc) [inline Consul.Test/SemaphoreTest.cs:458]:
```suggestion
            var lockWaitTimeMs = TimeSpan.FromSeconds(LockWaitTimeSeconds).TotalMilliseconds;
```

File: Consul.Test/SemaphoreTest.cs [hunk 3]
```diff
@@ -417,5 +417,54 @@ public async Task Cancelling_A_Token_When_Acquiring_A_Lock_Should_Throw_LockNotH
             await distributedLock2.Destroy();
             masterClient.Dispose();
         }
+
+        [Fact]
+        public async Task Cancelling_A_Token_When_Acquiring_A_Lock_Respects_The_Token()
+        {
+            const string key = "service/myApp/leader";
```
Comment (marcin-krystianc) [inline Consul.Test/SemaphoreTest.cs:424]: Using the same key name across several tests can affect other tests using same name of the key.
If one of the tests fails then then state of the agent will not be "clean" thus other tests can also fail in unexpected way.
For better test isolation I would suggest using some unique name for the key, e.g.:
```
var keyName = Path.GetTempFileName();
```

File: Consul.Test/SemaphoreTest.cs [hunk 4]
```diff
@@ -417,5 +417,54 @@ public async Task Cancelling_A_Token_When_Acquiring_A_Lock_Should_Throw_LockNotH
...
+            var distributedLock = _client.CreateLock(new LockOptions(key)
+            {
+                SessionTTL = TimeSpan.FromSeconds(DefaultSessionTTLSeconds),
+                LockWaitTime = TimeSpan.FromSeconds(LockWaitTimeSeconds)
+            });
+            var cancellationOperationTimer = new Stopwatch();
+            var cts = new CancellationTokenSource(TimeSpan.FromSeconds(1));
+
+            // Act
+            await distributedLock2.Acquire(); // Become "Master" with another instance first
```
[the remark this answers is not in the document]
Comment (mfkl) [reply Consul.Test/SemaphoreTest.cs]: Done

File: Consul.Test/SemaphoreTest.cs (hunk 2 above)
Comment (mfkl) [reply Consul.Test/SemaphoreTest.cs]: Done

File: Consul.Test/SemaphoreTest.cs (hunk 3 above)
Comment (mfkl) [reply Consul.Test/SemaphoreTest.cs]: Done

File: Consul.Test/SemaphoreTest.cs [hunk 5]
```diff
@@ -385,9 +386,9 @@ public async Task Passing_Cancelled_CancellationToken_Should_Throw_LockNotHeldEx
         }
 
         [Fact]
-        public async Task Cancelling_A_Token_When_Acquiring_A_Lock_Should_Throw_LockNotHeldException()
+        public async Task Cancelling_A_Token_When_Acquiring_A_Lock_Should_Throw_TaskCanceledException()
         {
-            const string keyName = "service/myApp/leader";
+            var keyName = Path.GetTempFileName();
```
Comment (marcin-krystianc) [inline Consul.Test/SemaphoreTest.cs:391]: Sorry, I meant `Path.GetRandomFileName()` 🤷

Comment (mfkl) [reply Consul.Test/SemaphoreTest.cs]: Done :-)

**Review by marcin-krystianc — APPROVED** [review]
LGTM 👍

**Pull request merged**